Add new TestRunner methods to Test Spec (#148)
microsoft/BotFramework-FunctionalTests: specs/TransciptTestRunner.md

- `_XYZ_Client`: specialized client that knows how to interact with a specific channel (it should be capable of generating channel OAuthTokens, handle responses coming from that specific channel, etc.)
- `ClientType` (enum): types are: DirectLine, Emulator, Teams

## TestRunner Class
### Methods

> - `RunTestAsync(string)`  Executes a single scenario based on the transcript passed to the method.   

Reviewer: I'm thinking we may want to rename this to `RunScriptAsync` or `RunTranscriptAsync` something different.

> - `RunTestAsync(string)`  Executes a single scenario based on the transcript passed to the method.   

Reviewer: I think we may also want to add a scriptType parameter here so the runner knows if it is an emulator transcript or a collection of testActions (TBD)

    * _Parameters_   
        * transcript `string`    The transcript that contains the scenario to execute.  
    * _Returns_  
A `Task`.    

    * Sample Usage   

> <img src="media/runtestasync1.png" width="700" border=1px>

Reviewer: Consider using code snippets rather than bitmaps so we can edit them
```csharp
var runner = new TestRunner(new TestClientFactory(ClientType.DirectLine).GetTestClient());
await runner.RunTestAsync(Path.Combine(_transcriptsFolder, transcript));
```

- `RunTestAsync(IEnumerable<string>) ` 
Executes a test based on a collection of transcripts, in sequence, in the order in which they exist in the collection. This method can be used to compose a complex scenario formed by a set of smaller scenarios defined in each transcript.  
    * _Parameters_   
        * transcripts `IEnumerable<string>` The collection of transcripts to execute.  
    * _Returns_  
A `Task`.   

    * Sample Usage    
<img src="media/runtestasync2.png" width="700" border=1px>
- `RunTestAsync(string, params string[])`  Executes a scenario utilizing a parameterized transcript.    
    * _Parameters_      
        * transcript `string`    The parameterized transcript that contains the scenario to execute.  
        * values `params string[]`    The parameters to be passed to the transcript.    
    * _Returns_  
A `Task`.  
    * Sample Usage  
<img src="media/runtestasync3.png" width="700" border=1px>
- `SendActivityAsync(Activity)`  Sends a single Activity to the test bot.    
    * _Parameters_    
        * activity `Microsoft.Bot.Schema.Activity`  The Activity that is sent to the bot.    
    * _Returns_
A `Task<bool>`

> - `ContinueWithAsync(string)` Executes a transcript as a continuation to the last scenario.

Reviewer: I don't think we need this, it may be easier to call RunTestAsync multiple times, based on what I see in the code right now.

    * _Parameters_    
        * transcript   `string`    The transcript that gets executed.

- `ContinueWithAsync(IEnumerable<string>)` Executes a collection of transcripts in sequence as a continuation to the last scenario.
    * _Parameters_    
        * transcripts   `IEnumerable<string>`    The transcripts that get executed.

- `ContinueWithAsync(string, params string[])` Executes a parameterized transcript as a continuation to the last scenario.
    * _Parameters_    
        * transcript   `string`    The transcript that gets executed.
        * params `params string[]`   The parameters to be passed to the transcript.

> - `GetActivitiesAsync()`  Gets a collection of activities returned by the bot.    

Reviewer: ```suggestion
- `GetNextReplyAsync()`  Gets the next reply returned by the bot.    
```

    * _Parameters_    
        * None.
    * _Returns_
A `Task<IEnumerable<Microsoft.Bot.Schema.Activity>>`   
* Complex Test Scenario Example   

> <img src="media/runtestasync4.png" width="700" border=1px>

Reviewer: ```csharp
var runner = new TestRunner(new TestClientFactory(ClientType.DirectLine).GetTestClient());

// Start with a transcript
await runner.RunTestAsync(Path.Combine(_transcriptsFolder, "BookFlight.transcript"));

// Continue by hand
await runner.SendActivityAsync(MessageFactory.Text("EchoSkillBot"));
await runner.AssertReplyAsync(activity =>
{
    Assert.Equal(ActivityTypes.Message, activity.Type);
    Assert.Equal("What skill would you like to call?", activity.Speak);
});
```

> 

Reviewer: Added a new AssertReplyAsync() to TestRunner that takes an action so we can do custom asserts on it
```csharp
var runner = new TestRunner(new TestClientFactory(ClientType.DirectLine).GetTestClient());
await runner.SendActivityAsync((Activity)Activity.CreateMessageActivity());
await runner.AssertReplyAsync(activity =>
{
    Assert.Equal(ActivityTypes.Message, activity.Type);
    Assert.Equal("Welcome to the Dialog Skill Prototype!", activity.Speak);
});
```
